backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java

```diff
@@ -428,6 +433,120 @@ public FutureExperimentRunDAO(
         .thenCompose(unused -> deleteExperimentRuns(runIds), executor);
   }
 
+  /**
+   * @param datasets : datasets for privilege check
+   * @param errorOut : Throw error while creation (true) otherwise we will keep it silent (false) at
+   *     fetch time
+   * @return {@link List} : accessible datasets
+   */
+  private InternalFuture<List<Artifact>> filterAndGetPrivilegedDatasetsOnly(
+      List<Artifact> datasets, boolean errorOut) {
+    Set<String> linkedDatasetVersionIds = new HashSet<>();
+    for (Artifact dataset : datasets) {
+      String datasetVersionId = dataset.getLinkedArtifactId();
+      if (!datasetVersionId.isEmpty()) {
+        linkedDatasetVersionIds.add(datasetVersionId);
+      }
+    }
+
+    // If linked dataset version ids are not exists in dataset then no need to check dataset
+    // authorization.
+    if (linkedDatasetVersionIds.isEmpty()) {
+      return InternalFuture.completedInternalFuture(datasets);
+    }
+    return jdbi.withHandle(
+            handle -> {
+              // We have dataset version as a commit and dataset as a repository in MDB so added
+              // commit_hash filter here
+              return handle
+                  .createQuery(
+                      " SELECT commit_hash, repository_id FROM repository_commit WHERE commit_hash IN (<linkedDatasetVersionIds>) ")
+                  .bindList("linkedDatasetVersionIds", linkedDatasetVersionIds)
+                  .map(
+                      (rs, ctx) ->
+                          new AbstractMap.SimpleEntry<>(
+                              rs.getString("commit_hash"), rs.getLong("repository_id")))
+                  .list();
+            })
+        .thenCompose(
+            datasetVersionDatasetList -> {
+              // Convert list of dataset and dataset version mapping into Map
+              Map<String, String> datasetVersionDatasetMap = new HashMap<>();
+              for (var entry : datasetVersionDatasetList) {
+                datasetVersionDatasetMap.put(entry.getKey(), String.valueOf(entry.getValue()));
+              }
+              return InternalFuture.completedInternalFuture(datasetVersionDatasetMap);
+            },
+            executor)
+        .thenCompose(
+            datasetVersionDatasetMap -> {
+              // Validate linked dataset in dataset version exists in database records if not then
+              // we will throw the error
```

Comment:
Under which scenarios would this happen? Sounds like a bug, no?

Reply:
No it is not a bug. So let say we have dataset and dataset version which is mapped with runs but after adding it as datasets  in ER, user will delete the datasets or that mapped dataset version at that time we are not deleting it from mapped ER so it will be still there with mapped ER table and while we are try to fetch, it will be not there so we should have to throw error.